File: visuals/src/Fable.Ripple/Builder.js

import { class_type } from "../../fable_modules/fable-library-js.5.13.0/Reflection.js";

/**
 * Computation expression for building computed signals. `let! ... and! ...` is
 * applicative (compiles to `map2`/`map3`); a lone `let!` is monadic (`bind`).
 * Every `Var` is a `Signal`, so a source or a derived signal bind alike.
 */
export class SignalBuilder {
    constructor() {
    }
}

export function SignalBuilder_$reflection() {
    return class_type("Fable.Ripple.SignalBuilder", undefined, SignalBuilder);
}

export function SignalBuilder_$ctor() {
    return new SignalBuilder();
}

export const SignalBuilderAuto_signal = SignalBuilder_$ctor();

File: src/Fable.Ripple/Builder.fs
namespace Fable.Ripple

/// Computation expression for building computed signals. `let! ... and! ...` is
/// applicative (compiles to `map2`/`map3`); a lone `let!` is monadic (`bind`).
/// Every `Var` is a `Signal`, so a source or a derived signal bind alike.
type SignalBuilder() =
    member inline _.Return(x: 'T) : Signal<'T> = Signal.constant x
    member inline _.ReturnFrom(s: Signal<'T>) : Signal<'T> = s

    member inline _.Bind(s: Signal<'a>, f: 'a -> Signal<'b>) : Signal<'b> = Signal.bind f s

    member inline _.BindReturn(s: Signal<'a>, f: 'a -> 'b) : Signal<'b> = Signal.map f s

    member inline _.MergeSources(a: Signal<'a>, b: Signal<'b>) : Signal<'a * 'b> =
        Signal.map2 (fun x y -> x, y) a b

[<AutoOpen>]
module SignalBuilderAuto =
    let signal = SignalBuilder()
